### PR TITLE
Add worker config table and view

### PR DESCRIPTION
**Pull request by talanknight**
Also the worker domain object contains a WorkerConfig object as well as methods for CanonicalAddress and LastConnectionUpdate which returns values based on a combination of worker and config values.

A new view has been added to get all tags and the config along with the worker in a single db query.

Target service has been refactored to use some of the new functions listed above and to try to make the worker filtering logic easier to follow.

### CODE REVIEW

File: internal/proto/controller/storage/servers/store/v1/worker.proto [hunk 1]
```diff
@@ -45,7 +45,7 @@ message Worker {
 // worker_id, key, value, and source.
 message WorkerTag {
   // worker_id is the public key that key of the worker this tag is for.
-  // @inject_tag: `gorm:"default:null"`
+  // @inject_tag: `gorm:"primary_key"`
```
Comment (irenarindos) [inline internal/proto/controller/storage/servers/store/v1/worker.proto:48]: I think the primary_key tag needs to also be added on key and value to represent the composite primary key for worker tags

File: internal/session/query.go [hunk 2]
```diff
@@ -324,8 +324,8 @@ where
 	closeConnectionsForDeadServersCte = `
    with
    dead_workers (worker_id, last_update_time) as (
-         select public_id, update_time
-           from server_worker
+         select worker_id, update_time
+           from server_worker_config
```
Comment (irenarindos) [inline internal/session/query.go:328]: Do we need to detect "dead" workers that don't have a config?

File: internal/servers/worker_status.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		tags = append(tags, &Tag{Key: t.Key, Value: t.Value})
+	}
+	cw := proto.Clone(w.WorkerStatus)
+	return &WorkerStatus{
+		WorkerStatus: cw.(*store.WorkerStatus),
+		Tags:         tags,
+	}
+}
+
+// TableName overrides the table name used by WorkerStatus to `server_worker_config`
```
Comment (irenarindos) [inline internal/servers/worker_status.go:44]:
```suggestion
// TableName overrides the table name used by WorkerStatus to `server_worker_status`
```

File: internal/servers/worker.go [hunk 4]
```diff
@@ -41,7 +48,142 @@ func (w *Worker) clone() *Worker {
...
+	}
+	cfg := NewWorkerStatus(a.PublicId, statusOptions...)
+	cfg.CreateTime = a.WorkerStatusCreateTime
+	cfg.UpdateTime = a.WorkerStatusUpdateTime
+
+	worker.ReportedStatus = cfg
+	return worker, nil
+}
+
+func tagsFromAggregatedTagString(ctx context.Context, s string) ([]*Tag, error) {
```
Comment (irenarindos) [inline internal/servers/worker.go:162]: Wondering if this function should be documented to explain the delimiters/ parsing.

**Review by irenarindos — APPROVED** [review]

File: internal/daemon/controller/handlers/targets/target_service_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	"testing"
+
+	"github.com/hashicorp/boundary/internal/servers"
+	"github.com/hashicorp/boundary/internal/types/scope"
+	pb "github.com/hashicorp/boundary/sdk/pbs/controller/api/resources/targets"
+	"github.com/hashicorp/go-bexpr"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (jimlambrt) [inline internal/daemon/controller/handlers/targets/target_service_test.go:14]: any chance you'd take a stab at implementing a basic unit test for `AuthorizeSession(...)`?

**Review by jimlambrt — CHANGES_REQUESTED** [review]
a few minor bits.  great work!

File: internal/servers/repository_worker.go [hunk 6]
```diff
@@ -25,20 +27,28 @@ func (r *Repository) listWorkersWithReader(ctx context.Context, reader db.Reader
 
 	var where string
 	if liveness > 0 {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/servers/repository_worker.go:35]: ty

File: internal/servers/repository_worker.go [hunk 7]
```diff
@@ -17,6 +18,7 @@ func (r *Repository) ListWorkers(ctx context.Context, opt ...Option) ([]*Worker,
 // listWorkersWithReader will return a listing of resources and honor the WithLimit option or the repo
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/servers/repository_worker.go]: ty

File: internal/daemon/controller/handlers/targets/target_service_test.go (hunk 5 above)
Comment (jimlambrt) [reply internal/daemon/controller/handlers/targets/target_service_test.go:14]: ty!

**Review by jimlambrt — APPROVED** [review]

File: internal/db/schema/migrations/oss/postgres/99/02_worker_controller_tables.up.sql [hunk 8]
```diff
@@ -58,6 +58,33 @@ create trigger worker_insert_time_column before insert on server_worker
...
+  worker_id wt_public_id primary key
+    constraint server_worker_fkey
+      references server_worker(public_id)
+      on delete cascade
+      on update cascade,
+  create_time wt_timestamp,
+  update_time wt_timestamp,
+  -- This is the calculated address that the worker reports it is reachable on.
+  address wt_network_address not null,
+  name wt_name
```
Comment (mgaffney) [inline internal/db/schema/migrations/oss/postgres/99/02_worker_controller_tables.up.sql:71]: What value is in the `name` column?

File: internal/session/query.go [hunk 9]
```diff
@@ -315,18 +315,25 @@ where
...
 	// along with their last update time and the number of connections closed on
-	// each.
+	// each.  If we do not know the last update time, we use the current time.
 	closeConnectionsForDeadServersCte = `
    with
    dead_workers (worker_id, last_update_time) as (
-         select public_id, update_time
-           from server_worker
-          where update_time < wt_sub_seconds_from_now(@grace_period_seconds)
+         select
```
Comment (jefferai) [inline internal/session/query.go:328]: The most minor of nits: you have a mix of spaces/tabs in this query

**Review by jefferai — APPROVED** [review]

**Pull request merged**